File: Code/Day02.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AoC.Code
{
    public class Day02
    {
        public static int Solve(IEnumerable<IEnumerable<int>> indata)
        {
            return indata.Select(row => row.Max() - row.Min()).Sum();
        }


        public static int Solve2(IEnumerable<IEnumerable<int>> indata)
        {
            return indata.Select(row => FindPair(row.ToList())).Sum();
        }


        private static int FindPair(List<int> indata)
        {
            for (int i = 0; i < indata.Count(); i++)
            {
                for (int j = 0; j < i; j++)
                {
                    var ii = indata[i];
                    var jj = indata[j];
                    if (ii % jj == 0)
                        return ii / jj;
                    if (jj % ii == 0)
                        return jj / ii;
                }
            }
            throw new Exception("No pair found in " + string.Join(", ", indata));
        }
    }
}

File: AoC.FSharp/Day02.fs
namespace AoC.FSharp

module Day02 =
    let solve (indata: int list list) = 
        let diffs = indata |> List.map(fun row ->
            let min = List.min(row)
            let max = List.max(row)
            max - min
        )
        List.sum diffs

    let solve2 (indata: int list list) = 
        let divisible x y =
            x <> y && x % y = 0
        
        let findPair (ints: int list) = 
            ints |> List.pick(fun i ->
                let j = ints |> List.tryFind(fun j -> divisible i j)
                match j with
                | Some j -> Some (i, j)
                | None -> None
            )

        let ratios = indata |> List.map(fun row ->
            let (a, b) = findPair row
            a / b
        )

        List.sum ratios